onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.cc

```diff
@@ -334,50 +337,77 @@ Status ProcessLogits(const OrtValue& logits,                                 //
     beam_state->remaining_scores = beam_state->remaining_scores.subspan(next_token_scores.size());
   }
 
-  // Apply top-k selection like the following:
-  //   next_token_scores = next_token_scores.view(batch_size, num_beams * vocab_size)
-  //   next_token_scores, next_tokens = torch.topk(next_token_scores, 2 * num_beams, dim=1, largest=True, sorted=True)
-  int64_t next_token_scores_dims[] = {batch_size, num_beams * vocab_size};
-  TensorShape next_token_scores_shape(&next_token_scores_dims[0], 2);
-  auto element_type = DataTypeImpl::GetType<float>();
-  OrtValue next_token_scores_value;
-  Tensor::InitOrtValue(element_type, next_token_scores_shape, next_token_scores.data(), allocator->Info(),
-                       next_token_scores_value);
-  const Tensor& input = next_token_scores_value.Get<Tensor>();
-
-  constexpr int axis = 1;
-  const unsigned top_k = static_cast<unsigned>(2 * num_beams);
-  constexpr bool largest = true;
-  constexpr bool sorted = true;  // results returned in sorted order.
-
-  std::unique_ptr<Tensor> topk_scores = Tensor::CreateDefault();
-  std::unique_ptr<Tensor> topk_indices = Tensor::CreateDefault();
-  ORT_RETURN_IF_ERROR(TopK(&input, axis, top_k, largest, sorted, allocator, stream, thread_pool,
-                           *topk_scores, *topk_indices));
-
+  if (num_beams <= 32) {
+    constexpr size_t max_parts_of_vocab = 128;
```

Comment:
Do we have tests that exercise both TopK paths ?

Reply:
I verified both path by test_generation.py. we don't have test for fp16 in test_generation.py because beam_search fp16 doesn't have exactly same result as transformer originally, so i didn't add the test code here.